### PR TITLE
Added autochip, autokick, move primitives

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Split MovePrimitive in FW into autochip, autokick, move primitives and use MoveManager to improve code reuse

### Testing Done
none...

### Resolved Issues
part of #1647

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware/app/primitives/move_manager.c [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+/**
+ * call from move_start to choose which wheel axis we will be
+ * using for preliminary rotation. The idea is to pick the wheel
+ * axis that will result in the minimum remaining rotation onto
+ * the bot's final destination angle.
+ *
+ * @param dx the global x position of the bot
+ * @param dy the global y position of the bot
+ * @param current_angle the current angle of the bot
+ * @param final_angle the final destination angle
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.c]: Missing javadoc [in/out]

File: src/firmware/app/primitives/move_manager.c [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+ */
+unsigned choose_wheel_axis(float dx, float dy, float current_angle, float final_angle);
+
+/**
+ * Calculates the rotation time, velocity, and acceleration to be stored
+ * in a PhysBot data container.
+ *
+ * @param pb The data container that has information about major axis time
+ * and will store the rotational information
+ * @param avel The current rotational velocity of the bot
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.c]: Missing javadoc [in/out]

File: src/firmware/app/primitives/move_manager.c [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+ */
+void plan_move_rotation(PhysBot* pb, float avel);
+
+/**
+ * builds an array that contains all of the axes perpendicular to
+ * each of the wheels on the bot.
+ *
+ * @param wheel_axes A pointer to the wheel_axes array to populate
+ * @param angle the current angle that the bot is facing
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.c]: Missing javadoc [in/out]

File: src/firmware/app/primitives/move_manager.c [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+ * in a PhysBot data container.
+ *
+ * @param pb The data container that has information about major axis time
+ * and will store the rotational information
+ * @param avel The current rotational velocity of the bot
+ */
+void plan_move_rotation(PhysBot* pb, float avel);
+
+/**
+ * builds an array that contains all of the axes perpendicular to
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.c]: Nit: Capitalize description to keep it consistent with the others.

File: src/firmware/app/primitives/move_manager.c [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+ * @param angle the current angle that the bot is facing
+ */
+void build_wheel_axes(float (*wheel_axes)[8], float angle)
+{
+    (*wheel_axes)[0] = angle + ANGLE_TO_FRONT_WHEELS - PI_2;
+    (*wheel_axes)[1] = angle + ANGLE_TO_FRONT_WHEELS + PI_2;
+    (*wheel_axes)[2] = angle - ANGLE_TO_FRONT_WHEELS - PI_2;
+    (*wheel_axes)[3] = angle - ANGLE_TO_FRONT_WHEELS + PI_2;
+    (*wheel_axes)[4] = angle + ANGLE_TO_BACK_WHEELS - PI_2;
+    (*wheel_axes)[5] = angle + ANGLE_TO_BACK_WHEELS - (3 * PI_2);
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.c]: Is there a reason for `3 * PI_2`? Anywhere we can comment on why these `PI_2` operations are so?

File: src/firmware/app/primitives/move_manager.c [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+ * If we are far enough away from our destination, then we should try
+ * rotating onto a wheel axis so that we can move faster. We should
+ * pick the wheel axis that minimizes the distance the bot will have
+ * to rotate to get to its destination angle after rotating onto an
+ * axis.
+ *
+ * @param optimal_wheel_axes_index The index of the wheel axes to rotate onto
+ * @param pb The data container that contains information about
+ * the direction the robot will move along.
+ * @param angle The angle that the robot is currently facing
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.c]: Missing javadoc [in/out]

File: src/firmware/app/primitives/move_manager.c [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    assert(!isnan(state->minor_vec[1]));
+    const FirmwareRobot_t* robot = app_firmware_world_getRobot(world);
+
+    PhysBot pb =
+        app_physbot_create(robot, state->destination, state->major_vec, state->minor_vec);
+
+    // choose a wheel axis to rotate onto
+    // TODO: try to make this less jittery
+    //    choose_rotation_destination(state->optimial_wheel_axes_index, &pb,
+    //    current_states.angle);
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.c]: Outstanding TODO comment.

File: src/firmware/app/primitives/move_manager.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+#include "firmware/app/world/firmware_world.h"
+#include "shared/proto/primitive.nanopb.h"
+
+/**
+ * Start moving in the given world
+ * @param void_state_ptr [in] A pointer to this primitives state, as allocated by the
+ *                            `create_state` function
+ * @param world [in] The world this primitive is running in.
+ * @param move_position_params Move Position Params
+ * @param final_angle The final angle
```
Comment (kelvinkoon) [inline src/firmware/app/primitives/move_manager.h]: Missing javadoc [in/out]

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Nice work! :100:

File: src/firmware/app/primitives/move_manager.c (hunk 1 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: with no [in/out] it should assumed that they are all [in]

File: src/firmware/app/primitives/move_manager.c (hunk 5 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: this was copied from the MovePrimitive, so fixing these things are out of scope

File: src/firmware/app/primitives/move_manager.c (hunk 7 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: this was copied from the MovePrimitive, so fixing these things are out of scope

File: src/firmware/app/primitives/move_manager.c (hunk 4 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: fixed

**Review by kelvinkoon — DISMISSED** [review]

File: src/firmware/app/primitives/move_manager.c (hunk 1 above)
Comment (MathewMacDougall) [reply src/firmware/app/primitives/move_manager.c]: lets have further discussion in #1650

File: src/firmware/app/primitives/move_new_primitive.c [hunk 9]
```diff
@@ -0,0 +1,46 @@
+#include "firmware/app/primitives/move_new_primitive.h"
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "firmware/app/primitives/move_manager.h"
+
+void app_move_primitive_start(MovePrimitiveMsg prim_msg, void *void_state_ptr,
```
Comment (MathewMacDougall) [inline src/firmware/app/primitives/move_new_primitive.c]: Are we not deleting the old move primitive as part of this PR? Any reason to temporarily keep both around?

File: src/firmware/app/primitives/move_manager.c [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+// The minimum distance away from our destination that we must be if we
+// are going to rotate the bot onto its wheel axis
+// 2 * P_PI * ROBOT_RADIUS = robot circumference, which is approximately
+// how far the bot would have to turn for one full rotation, so we
+// set it a litle larger than that.
+const float APPROACH_LIMIT = 3 * P_PI * ROBOT_RADIUS;
+
+const float PI_2 = P_PI / 2.0f;
+
+typedef struct MoveManagerState
```
Comment (MathewMacDougall) [inline src/firmware/app/primitives/move_manager.c]: I like the idea of the MoveManager to help reduce duplicated code. Just wondering if we can think of a better name than "manager" since I wouldn't initially assume it actually does most of the movement.

I understand why we can't just have "move with autochip" call the regular MovePrimitive (MovePrimitive disables autokick/chip for safetyand correctness), so just feel like a slightly different name would be helpful here.

random ideas:
* CoreMove / MoveCore
* MoveHelper

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looks great overall

File: src/firmware/app/primitives/move_new_primitive.c (hunk 9 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_new_primitive.c]: until https://github.com/UBC-Thunderbots/Software/issues/1643 is done, we're still using the old primitiveparams, so to avoid name conflicts, I'm making a new moveprimitive that uses the new primitivemsg that can be swapped in later

File: src/firmware/app/primitives/move_manager.c (hunk 10 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: changed to MoveHelper

Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: The thought is that MoveHelper will soon be replaced by trajectory planner and movepositionparams will be replaced by a trajectory planner proto

File: src/firmware/app/primitives/move_new_primitive.c (hunk 9 above)
Comment (garethellis0) [reply src/firmware/app/primitives/move_new_primitive.c]: Why not just move over one primitive at a time? :thinking:

File: src/firmware/app/primitives/move_manager.h [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+ */
+void app_move_manager_start(void *void_state_ptr, FirmwareWorld_t *world,
+                            MovePositionParams move_position_params, float final_angle);
```
Comment (garethellis0) [inline src/firmware/app/primitives/move_manager.h]: Why not call this `move_primitive` instead? It effectively acts as one, no?

File: src/firmware/app/primitives/autochip_move_primitive.c [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+                                       void* void_state_ptr, FirmwareWorld_t* world)
+{
+    app_move_manager_start(void_state_ptr, world, prim_msg.position_params,
+                           prim_msg.final_angle.radians);
+
+    const FirmwareRobot_t* robot = app_firmware_world_getRobot(world);
+
+    Dribbler_t* dribbler = app_firmware_robot_getDribbler(robot);
+    Chicker_t* chicker   = app_firmware_robot_getChicker(robot);
+    app_dribbler_setSpeed(dribbler, (uint32_t)prim_msg.dribbler_speed_rpm);
```
Comment (garethellis0) [inline src/firmware/app/primitives/autochip_move_primitive.c:19]: Should we spin the dribbler in the `autochip` primitive? Seems like we could damage the roller more by dribbling while chipping.

File: src/firmware/app/primitives/move_manager.h (hunk 11 above)
Comment (garethellis0) [inline src/firmware/app/primitives/move_manager.h]: Why not call this "move_primitive" instead? That's effectively what this is, no?

File: src/firmware/app/primitives/autochip_move_primitive.c [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    app_chicker_enableAutochip(chicker, prim_msg.chip_distance_meters);
+}
+
+static void autochip_move_end(void* void_state_ptr, FirmwareWorld_t* world)
```
Comment (garethellis0) [inline src/firmware/app/primitives/autochip_move_primitive.c:23]: missing `app_` prefix here

File: src/firmware/app/primitives/autochip_move_primitive.c [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+    Chicker_t* chicker = app_firmware_robot_getChicker(robot);
+    app_chicker_disableAutochip(chicker);
+    app_chicker_disableAutokick(chicker);
+
+    Dribbler_t* dribbler = app_firmware_robot_getDribbler(robot);
+    app_dribbler_setSpeed(dribbler, 0);
+}
+
+static void autochip_move_tick(void* void_state_ptr, FirmwareWorld_t* world)
```
Comment (garethellis0) [inline src/firmware/app/primitives/autochip_move_primitive.c:36]: Missing `app_` prefix here

Comment (garethellis0) [reply src/firmware/app/primitives/autochip_move_primitive.c:36]: Similar for a few functions below. Could you do a pass through all the functions in this PR?

File: src/firmware/app/primitives/move_manager.c (hunk 10 above)
Comment (garethellis0) [reply src/firmware/app/primitives/move_manager.c]: I'm a bit confused why we don't just have a `move_primitive` that moves without autokick or autochip, then the `move_with_autokick` and `move_with_autochip` primitives can just use the `move_primitive`.

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/firmware/app/primitives/move_new_primitive.c (hunk 9 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_new_primitive.c]: yeah I can delete move_new_primitive

File: src/firmware/app/primitives/autochip_move_primitive.c (hunk 13 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/autochip_move_primitive.c:23]: will be done here: https://github.com/UBC-Thunderbots/Software/issues/1631

File: src/firmware/app/primitives/move_manager.c (hunk 10 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: move_primitive would take MovePrimitiveMsg so it would require move_with_autokick to convert its AutokickMovePrimitiveMsg into MovePrimitiveMsg

File: src/firmware/app/primitives/move_manager.h (hunk 11 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.h]: yes but it only handles the movement part of the move_primitive, not the dribbling

File: src/firmware/app/primitives/autochip_move_primitive.c (hunk 12 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/autochip_move_primitive.c:19]: hm, how are we supposed to do backspin chipping then?

File: src/firmware/app/primitives/move_manager.h (hunk 11 above)
Comment (garethellis0) [reply src/firmware/app/primitives/move_manager.h]: Sorry, still a bit confused here. Seems like semantically this could just be "move without dribbling or autokicking or autochipping" and then the other primitives could just build off this one?

File: src/firmware/app/primitives/autochip_move_primitive.c (hunk 12 above)
Comment (garethellis0) [reply src/firmware/app/primitives/autochip_move_primitive.c:19]: @MathewMacDougall for his input. Not a blocking comment, just a thought. You're right that we would need this for backspin

File: src/firmware/app/primitives/move_manager.c (hunk 10 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.c]: Gonna run into the same issue as https://github.com/UBC-Thunderbots/Software/pull/1645#discussion_r464014504 if I turn it into a MovePrimitive and we can separate "moving" from "moving and dribbling", so not changing

File: src/firmware/app/primitives/move_manager.h (hunk 11 above)
Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.h]: Gonna run into the same issue as https://github.com/UBC-Thunderbots/Software/pull/1645#discussion_r464014504 so I don't think it's worth it

Comment (jonathanlew) [reply src/firmware/app/primitives/move_manager.h]: I think it'll be nuked and replaced by trajectory follower or something in https://github.com/UBC-Thunderbots/Software/issues/1654, so I'll leave it as move_helper for now

File: src/firmware/app/primitives/move_manager.c (hunk 10 above)
Comment (MathewMacDougall) [reply src/firmware/app/primitives/move_manager.c]: Another reason it wouldn't work @garethellis0 is that I assumed the MovePrimitive would disable autokick and autochip (because that's more correct and would prevent issues if shoot didn't disable it or something). However because the MovePrimitive would be called by MoveWithAutokick, it would disable autokick after it was set, so we'd never be able to move with autokick on.

File: src/firmware/app/primitives/autochip_move_primitive.c (hunk 12 above)
Comment (MathewMacDougall) [reply src/firmware/app/primitives/autochip_move_primitive.c:19]: 1. This code is not backspin chipping (as I think of it). Because autochip is enabled there's no time for the ball to build up any backspin so there'd be no noticeable effect on the ball when it lands

2. Unsure about how much damage this may cause. Using the dribbler is a reasonable idea to try get more _consistent_ chips because hopefully it will put the ball in a more consistent position so it makes better contact with the chipper (inconsistency typically comes from the ball being too far away or too close). This may require testing in different scenarios to see if it's worth it. In most cases where we'd use this there should be relative velocity between the ball and robot on contact (robot moves into ball, or ball moves into robot). This _should_ result in better contact without the need for the dribbler so perhaps it's not needed. Also depending on how the dribbler behaves and when the breakbeam fires, a running dribbler may knock the ball away if it doesn't catch it properly for some reason.

tl;dr we need to do tests to check for dribbler damage, and consistency gains / losses.

**Review by MathewMacDougall — COMMENTED** [review]
Replied to a few comments but nothing major to add. Seems like there's still a bit of consufion around the MoveHelper so maybe worth a call or quick discussion on Slack @jonathanlew @garethellis0

**Review by garethellis0 — APPROVED** [review]
I'm fine with `move_manager` for now since I'm going to rewrite this anyhow once I migrate to using the trajectory planner.

LGTM :+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
